Further improvements on textarea

ilpav12 opened the pull request:
Changed `x-on:load.window` to `x-init` and added `wire:ignore` after discussion on PR #8177 

- [x] Changes have been thoroughly tested to not break existing functionality.
- [x] New functionality has been documented or existing documentation has been updated to reflect changes.
- [x] Visual changes are explained in the PR description using a screenshot/recording of before and after.

ilpav12 committed:
changed x-on:load.window to x-init and added wire:ignore

ilpav12 committed:
Merge branch '3.x' into fix/textarea-autosize

zepfietje commented on packages/forms/resources/views/components/textarea.blade.php:
```diff
@@ -13,9 +13,10 @@
+            x-init="render()"
             x-on:input="render()"
             x-on:resize.window="render()"
+            wire:ignore
```
Doesn't this have negative side effects?
I think this breaks error state?

ilpav12 replied:
@qrasa1 suggested this. I am not aware of what the side effects are

qrasa1 replied:
I mean... it shouldn't? I'm not a Livewire expert by any means, but if it did have some negative side effects, wouldn't that be mentioned in the [docs](https://laravel-livewire.com/docs/2.x/alpine-js#ignoring-dom-changes)?

Edit: I think you mean that the errors will stay on the page after saving? It shouldn't happen since the errors aren't added inside the `<textarea>`. I did test the required error and it disappeared after saving.

danharrin replied:
It will break conditionally disabling the field, we can't use wire:ignore here

zepfietje replied:
Also the error styling wouldn't be rendered.

qrasa1 replied:
I see... here is one possible workaround:
`textarea.blade.php` (note the `h-full` near the end):
```blade
    <div
        @if ($shouldAutosize = $shouldAutosize())
            ax-load
            ax-load-src="{{ \Filament\Support\Facades\FilamentAsset::getAlpineComponentSrc('textarea', 'filament/forms') }}"
            x-data="textareaFormComponent({ initialHeight: @js($initialHeight) })"
            x-ignore
            wire:ignore.self
            x-on:input="render()"
            x-on:resize.window="render()"
            style="height: {{ $initialHeight }}rem"
        @endif
    >
        <textarea
            @if($shouldAutosize)
                {{ $getExtraAlpineAttributeBag() }}
            @endif
            {{
                $attributes
                    ->merge([
                        'autocomplete' => $getAutocomplete(),
                        'autofocus' => $isAutofocused(),
                        'cols' => $getCols(),
                        'disabled' => $isDisabled(),
                        'id' => $getId(),
                        'maxlength' => (! $isConcealed) ? $getMaxLength() : null,
                        'minlength' => (! $isConcealed) ? $getMinLength() : null,
                        'placeholder' => $getPlaceholder(),
                        'readonly' => $isReadOnly(),
                        'required' => $isRequired() && (! $isConcealed),
                        'rows' => $rows,
                        $applyStateBindingModifiers('wire:model') => $statePath,
                    ], escape: false)
                    ->merge($getExtraAttributes(), escape: false)
                    ->merge($getExtraInputAttributes(), escape: false)
                    ->class([
                        'fi-fo-textarea block w-full h-full rounded-lg border-none bg-white px-3 py-1.5 text-base text-gray-950 shadow-sm outline-none ring-1 transition duration-75 placeholder:text-gray-400 focus:ring-2 disabled:bg-gray-50 disabled:text-gray-500 disabled:[-webkit-text-fill-color:theme(colors.gray.500)] disabled:placeholder:[-webkit-text-fill-color:theme(colors.gray.400)] dark:bg-white/5 dark:text-white dark:placeholder:text-gray-500 dark:disabled:bg-transparent dark:disabled:text-gray-400 dark:disabled:[-webkit-text-fill-color:theme(colors.gray.400)] dark:disabled:placeholder:[-webkit-text-fill-color:theme(colors.gray.500)] sm:text-sm sm:leading-6',
                        'ring-gray-950/10 focus:ring-primary-600 dark:ring-white/20 dark:focus:ring-primary-500 dark:disabled:ring-white/10' => ! $errors->has($statePath),
                        'ring-danger-600 focus:ring-danger-600 dark:ring-danger-500 dark:focus:ring-danger-500' => $errors->has($statePath),
                    ])
            }}
        ></textarea>
    </div>
``` 
`textarea.js`:
```javascript
export default function textareaFormComponent({ initialHeight }) {
    return {
        init: function () {
          setTimeout(() => this.render())
        },
        render: function () {
            if (this.$el.children[0].scrollHeight > 0) {
                this.$el.style.height = initialHeight + 'rem'
                this.$el.style.height = this.$el.children[0].scrollHeight + 'px'
            }
        },
    }
}
``` 
I thought that the `$getExtraAlpineAttributeBag()` belongs to the text area so I added there.

ilpav12 replied:
Ok thank you for the explanation of the drawbacks. Is there any event we can listen for to call the `render()` when you save the edit form?

ilpav12 replied:
One solution I have found is to add `wire:key="{{ rand() }}"` which forces Livewire to destroy the old components and to create a new one doing again the `init()`. This does not fix the problem with the modal on second opening.

danharrin replied:
We should not do that, its bad practice to use a random key

zepfietje commented on packages/forms/resources/views/components/textarea.blade.php:
```diff
@@ -13,9 +13,10 @@
             x-data="textareaFormComponent({ initialHeight: @js($initialHeight) })"
             x-ignore
-            x-on:load.window="render()"
+            x-init="render()"
```
Could you move this to `init()` inside the JS file maybe?

ilpav12 replied:
Done in commit https://github.com/filamentphp/filament/pull/8193/commits/ef67b59ea81097a7345ba20d356c1e83a5a8d702

ilpav12 committed:
Moved init render to .js file

ilpav12 committed:
Replaced `wire:ignore` with wire:key="{{ rand() }}" to fix height reset on Livewire refresh

danharrin commented:
Please remove the random key, we shouldnt do that

ilpav12 commented:
Ok do you have any suggestion on how to fix the problem?

ilpav12 committed:
Rollback on random key

zepfietje commented:
Just a note: GitHub doesn't even seem to render auto-growing textareas at the full content height on page load, only after you start typing. Makes me think what's actually the best UX in this case.

ilpav12 commented:
Ok interesting point. I think that the possibility to immediately have a glance to the full content of a textarea without manually expanding each of them is very convenient, but if you think it is unnecessary we can rollback all the changes and update only on input.

qrasa1 commented:
We could use something like [IntersectionObserver](https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver#examples) .
While this could work, it's ultimately up to you to decide whether it's worth it or not. Using markdown editor with no toolbar buttons might be a good alternative.

ilpav12 committed:
Add intersection observer to textarea component

An intersection observer provided by the @alpinejs/intersect plugin has been added to the textarea form component for better fixing wrong height on modal interactions.

ilpav12 commented:
Little update: I have added an intersection observer as suggest by @qrasa1 on https://github.com/filamentphp/filament/pull/8193/commits/285b38bf4577275b376e882b85dba9f5b704e446. This indeed fixes the behavior with modal. I am still looking for a fix on a full page save.

ilpav12 commented:
Summarizing the updates as of now, the following cases are fixed:
- initial page load (in dedicated `init()` function instead of  `x-on:load.window` as it is live now)
- browser resize (with` x-on:resize.window="render()"`) 
- modal opening (with `x-intersect.once="render()"`).
Instead, in case of page refresh with Livewire the height is reset. The possible solution are:
- `wire:ignore` but it breaks the conditional disabling
- `wire:key="{{ rand() }}"` but it not a best practice
- listen to a global Livewire page refresh event but I did not manage to find one.

danharrin committed:
wip

danharrin committed:
Merge branch '3.x' into pr/8193

danharrin committed:
wip

danharrin commented:
Thanks
